sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala

```diff
@@ -208,3 +208,161 @@ object ExtractPythonUDFFromJoinCondition extends Rule[LogicalPlan] with Predicat
       }
   }
 }
+
+sealed abstract class BuildSide
+
+case object BuildRight extends BuildSide
+
+case object BuildLeft extends BuildSide
+
+trait JoinSelectionHelper {
```

Comment:
We need to pull out this functionality as a helper trait? For example, `PredicateHelper` is used in many places, but this helper trait is only used for `JoinSelection `?

Reply:
@maropu That's true that currently there are no other uses of this trait, but there are potentially other rules that can use the join selection logic at optimization time, so this refactoring will open up the opportunity for more optimization rules.
Also, this cleans up the JoinSelection rule quite a bit, and IMO the logic for determining the type of the join is more clear now.

Reply:
okay, that sounds fine to me. btw, if the helper is more general as you said, how about `JoinSelectionHelper` -> `JoinPlanHelper `?

Reply:
Also, could you add some tests for this new helper?